pkg/rbac/sharedData.go

```diff
@@ -32,116 +39,164 @@ type resource struct {
 	kind     string
 }
 
-func (cache *Cache) ClusterScopedResources(ctx context.Context) ([]resource, error) {
-	clusterScoped, err := cache.shared.getClusterScopedResources(cache, ctx)
-	return clusterScoped, err
+var clusterVersionGvr = schema.GroupVersionResource{
+	Group:    "cluster.open-cluster-management.io",
+	Version:  "v1",
+	Resource: "managedclusters",
 }
 
-func (shared *SharedData) getClusterScopedResources(cache *Cache, ctx context.Context) ([]resource,
-	error) {
+func (cache *Cache) PopulateSharedCache(ctx context.Context) error {
 
-	// lock to prevent checking more than one at a time and check if cluster scoped resources already in cache
-	shared.csLock.Lock()
-	defer shared.csLock.Unlock()
-	if shared.csResources == nil ||
-		time.Now().After(shared.csUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL)*time.Millisecond)) {
-
-		// if not in cache query database
-		klog.V(6).Info("Querying database for cluster-scoped resources.")
-		shared.csErr = nil // Clear previous errors.
-
-		// Building query to get cluster scoped resources
-		// Original query: "SELECT DISTINCT(data->>apigroup, data->>kind) FROM search.resources WHERE
-		// cluster='local-cluster' AND namespace=NULL"
-
-		schemaTable := goqu.S("search").Table("resources")
-		ds := goqu.From(schemaTable)
-		query, _, err := ds.SelectDistinct(goqu.COALESCE(goqu.L(`"data"->>'apigroup'`), "").As("apigroup"),
-			goqu.COALESCE(goqu.L(`"data"->>'kind_plural'`), "").As("kind")).
-			Where(goqu.L(`"cluster"::TEXT = 'local-cluster'`), goqu.L(`"data"->>'namespace'`).IsNull()).ToSQL()
-		if err != nil {
-			klog.Errorf("Error creating query [%s]. Error: [%+v]", query, err)
-			shared.csErr = err
-			shared.csResources = []resource{}
-			return shared.csResources, shared.csErr
-		}
+	if sharedCacheValid(&cache.shared) { //if all cache is valid we use cache data
+		klog.V(5).Info("Using shared data from cache.")
+		return nil
+	} else { //get data and cache
 
-		rows, queryerr := cache.pool.Query(ctx, query)
-		if queryerr != nil {
-			klog.Errorf("Error resolving query [%s]. Error: [%+v]", query, queryerr.Error())
-			shared.csErr = queryerr
-			shared.csResources = []resource{}
-			return shared.csResources, shared.csErr
+		// get all cluster-scoped resources and cache in shared.csResources
+		err := cache.shared.getClusterScopedResources(cache, ctx)
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved cluster scoped resources!")
 		}
+		// get all namespaces in cluster and cache in shared.namespaces.
+		err = cache.shared.GetSharedNamespaces(cache, ctx)
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved shared namespaces!")
+		}
+		// get all managed clustsers in cache
+		err = cache.shared.GetSharedManagedClusters(cache, ctx)
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved managed clusters!")
+		}
+
+		return err
```

Comment:
These 3 actions should happen at the same time.
Here's an example of how to implement it using sync.WaitGroup.  https://gobyexample.com/waitgroups

Reply:
Okay, as discussed, will note this as optimization step in next user story.